File: rule/dev_launch_wrapper/cpp/plugin/type_plugin.cpp
#include "type_plugin.h"

#include <fstream>
#include <string_view>

static void write_file(const std::string& path, const std::string_view& content) {
    std::ofstream stream;
    stream.exceptions(std::ios::failbit | std::ios::badbit);
    stream.open(path);
    stream << content;
}

bool TypePlugin::process_arg(const std::string& arg, PluginContext& pctx, ArgumentContext& actx) {
    constexpr std::string_view prefix = "-Ddev.launch.type=";
    if (arg.rfind(prefix, 0) == 0) {
        type = arg.substr(prefix.size());
        return true;
    }
    return false;
}

void TypePlugin::finalize(PluginContext& pctx) {
    if (type == "server") {
        if (!std::filesystem::exists("server.properties")) {
            write_file("server.properties", "online-mode=false\n");
        }
        write_file("eula.txt", "eula=true\n");
        pctx.append_program_arg("--nogui");
    } else if (type.empty() || type == "client") {
        write_file("allowed_symlinks.txt", "[regex].*\n");
    }
}

File: rule/dev_launch_wrapper/java/plugin/TypePlugin.java
package top.fifthlight.fabazel.devlaunchwrapper.plugin;

import top.fifthlight.fabazel.devlaunchwrapper.DevLaunchContext;
import top.fifthlight.fabazel.devlaunchwrapper.DevLaunchPlugin;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public class TypePlugin implements DevLaunchPlugin {
    @Override
    public int priority() {
        return 700;
    }

    @Override
    public void load(DevLaunchContext context) throws IOException {
        String type = DevLaunchPlugin.property("type", "client");
        switch (type) {
            case "client": {
                Path allowSymlinksPath = Paths.get("allowed_symlinks.txt");
                DevLaunchPlugin.writeString(allowSymlinksPath, "[regex].*\n");
                break;
            }
            case "server": {
                Path serverPropertiesPath = Paths.get("server.properties");
                if (!Files.exists(serverPropertiesPath)) {
                    DevLaunchPlugin.writeString(serverPropertiesPath, "online-mode=false\n");
                }
                context.addArg("--nogui");
                Path eulaPath = Paths.get("eula.txt");
                DevLaunchPlugin.writeString(eulaPath, "eula=true\n");
                break;
            }
        }
    }
}
